### PR TITLE
http3: respecting header number limits

### PR DESCRIPTION
**Pull request by alyssawilk**
Respecting upstream and downstream caps on number of headers for HTTP/3

Risk Level: n/a (http/3)
Testing: turned up integration tests
Docs Changes: n/a
Release Notes: n/a
#14829  among others.

### CODE REVIEW

File: test/integration/protocol_integration_test.cc [hunk 1]
```diff
@@ -1600,7 +1600,6 @@ TEST_P(DownstreamProtocolIntegrationTest, ManyRequestHeadersAccepted) {
 TEST_P(DownstreamProtocolIntegrationTest, ManyRequestTrailersRejected) {
   // QUICHE doesn't limit number of headers.
   EXCLUDE_DOWNSTREAM_HTTP3
```
Comment (danzh2010) [inline test/integration/protocol_integration_test.cc:1602]: The upstream http3 test failure is irrelevant to this fix right? I doubt current fix works for trailers.

File: test/integration/multiplexed_upstream_integration_test.cc [hunk 2]
```diff
@@ -47,12 +47,10 @@ TEST_P(Http2UpstreamIntegrationTest, RouterHeaderOnlyRequestAndResponseNoBuffer)
 }
 
 TEST_P(Http2UpstreamIntegrationTest, RouterUpstreamDisconnectBeforeRequestcomplete) {
-  EXCLUDE_UPSTREAM_HTTP3; // Close loop.
```
Comment (danzh2010) [inline test/integration/multiplexed_upstream_integration_test.cc:50]: How are these tests related to max headers count?

File: source/common/quic/envoy_quic_utils.h [hunk 3]
```diff
@@ -45,16 +45,26 @@ class HeaderValidator {
   validateHeader(const std::string& header_name, absl::string_view header_value) = 0;
 };
 
+static constexpr absl::string_view invalid_underscore = "http3.unexpected_underscore";
+static constexpr absl::string_view too_many_http_headers = "http3.too_many_http_headers";
```
Comment (danzh2010) [inline source/common/quic/envoy_quic_utils.h:49]: Maybe add/reuse those already defined in Http3ResponseCodeDetailValues? You may need to move it around.

File: source/common/quic/envoy_quic_utils.h [hunk 4]
```diff
@@ -45,16 +45,26 @@ class HeaderValidator {
...
+    if (max_headers_allowed == 0) {
+      details = too_many_http_headers;
+      return nullptr;
+    }
+    max_headers_allowed--;
     Http::HeaderUtility::HeaderValidationResult result =
         validator.validateHeader(entry.first, entry.second);
     switch (result) {
     case Http::HeaderUtility::HeaderValidationResult::REJECT:
+      details = invalid_underscore;
```
Comment (danzh2010) [inline source/common/quic/envoy_quic_utils.h:67]: This details is ready assigned in server stream's validateHeader().

File: source/common/quic/envoy_quic_utils.h [hunk 5]
```diff
@@ -45,16 +45,26 @@ class HeaderValidator {
...
+static constexpr absl::string_view invalid_underscore = "http3.unexpected_underscore";
+static constexpr absl::string_view too_many_http_headers = "http3.too_many_http_headers";
+
 // The returned header map has all keys in lower case.
 template <class T>
-std::unique_ptr<T> quicHeadersToEnvoyHeaders(const quic::QuicHeaderList& header_list,
-                                             HeaderValidator& validator) {
+std::unique_ptr<T>
+quicHeadersToEnvoyHeaders(const quic::QuicHeaderList& header_list, HeaderValidator& validator,
+                          uint32_t max_headers_allowed, absl::string_view& details) {
```
Comment (danzh2010) [inline source/common/quic/envoy_quic_utils.h:55]: Is it because QuicHeaderList doesn't have expose size() interface that we need max_headers_allowed to be passed into this function?

I slightly prefer checking it in QUICHE and have something like OnHeadersTooLarge() or simply add QuicHeaderList::size() instead of iterating over the header list again.

File: source/common/quic/envoy_quic_client_stream.cc [hunk 6]
```diff
@@ -138,16 +139,18 @@ void EnvoyQuicClientStream::OnInitialHeadersComplete(bool fin, size_t frame_len,
     end_stream_decoded_ = true;
   }
   std::unique_ptr<Http::ResponseHeaderMapImpl> headers =
-      quicHeadersToEnvoyHeaders<Http::ResponseHeaderMapImpl>(header_list, *this);
+      quicHeadersToEnvoyHeaders<Http::ResponseHeaderMapImpl>(
+          header_list, *this, filterManagerConnection()->maxIncomingHeadersCount(), details_);
   if (headers == nullptr) {
-    onStreamError(close_connection_upon_invalid_header_);
+    onStreamError(close_connection_upon_invalid_header_, quic::QUIC_STREAM_INTERNAL_ERROR);
```
Comment (danzh2010) [inline source/common/quic/envoy_quic_client_stream.cc:145]: This is not negotiated in initial SETTINGS. I think it should be QUIC_BAD_APPLICATION_PAYLOAD or at least something else other than QUIC_STREAM_INTERNAL_ERROR?

**Review by danzh2010 — COMMENTED** [review]
Thanks for fixing more tests! A few thoughts on the details.

File: source/common/quic/envoy_quic_utils.h (hunk 5 above)
Comment (danzh2010) [reply source/common/quic/envoy_quic_utils.h:73]: SGTM.

File: source/common/quic/envoy_quic_client_stream.cc (hunk 6 above)
Comment (danzh2010) [reply source/common/quic/envoy_quic_client_stream.cc:145]: How about QUIC_STREAM_EXCESSIVE_LOAD ?

File: source/common/quic/envoy_quic_server_stream.cc [hunk 7]
```diff
@@ -249,6 +250,11 @@ void EnvoyQuicServerStream::maybeDecodeTrailers() {
   if (sequencer()->IsClosed() && !FinishedReadingTrailers()) {
     // Only decode trailers after finishing decoding body.
     end_stream_decoded_ = true;
+    if (received_trailers().size() > filterManagerConnection()->maxIncomingHeadersCount()) {
+      details_ = Http3ResponseCodeDetailValues::too_many_trailers;
+      onStreamError(close_connection_upon_invalid_header_);
+      return;
+    }
```
Comment (danzh2010) [inline source/common/quic/envoy_quic_server_stream.cc:257]: Same for client stream?

File: test/integration/protocol_integration_test.cc [hunk 8]
```diff
@@ -1622,13 +1617,12 @@ TEST_P(DownstreamProtocolIntegrationTest, ManyRequestTrailersRejected) {
     EXPECT_TRUE(response->complete());
     EXPECT_EQ("431", response->headers().getStatusValue());
   } else {
-    ASSERT_TRUE(response->waitForReset());
```
Comment (danzh2010) [inline test/integration/protocol_integration_test.cc:1625]: should this be reverted?

File: source/common/quic/envoy_quic_utils.h [hunk 9]
```diff
@@ -31,6 +31,27 @@
...
+  static constexpr absl::string_view invalid_http_header = "http3.invalid_header_field";
+  // The size of headers (or trailers) exceeded the configured limits.
+  static constexpr absl::string_view headers_too_large = "http3.headers_too_large";
+  // Envoy was configured to drop requests with header keys beginning with underscores.
+  static constexpr absl::string_view invalid_underscore = "http3.unexpected_underscore";
+  // The peer refused the stream.
+  static constexpr absl::string_view remote_refused = "http3.remote_refuse";
+  // The peer reset the stream.
+  static constexpr absl::string_view remote_reset = "http3.remote_reset";
+  // Too many trailers were sent
```
Comment (snowp) [inline source/common/quic/envoy_quic_utils.h:49]: period

**Review by snowp — COMMENTED** [review]
Seems right to me, just one tiny nit

**Comment by alyssawilk** [comment]
Looks like an asan flake in a new test
/wait

**Review by danzh2010 — APPROVED** [review]

**Comment by alyssawilk** [comment]
woot! period added, merge conflicts sorted, test flake fixed :-)

**Review by snowp — APPROVED** [review]
Thanks!
